### ISSUE TITLE
PS256 algorithm support?

### ISSUE THREAD

**Issue by FractalizeR** [lcobucci/jwt#1074]
Hello.

PS256 is listed as unsupported on jwt.io for lcobucci. I need this algo for authenticating with one cloud operator. Any plans for it to be supported?

**Comment by Ocramius** [comment]
I glanced over it a bit, and it seems like an implementation -only detail that doesn't need external dependencies.

https://auth0.com/docs/get-started/applications/signing-algorithms

> Unlike RS256, the same JWT header and payload will generate a different signature each time.

If that's true (I didn't look further), then I encourage starting a patch.

I'm unsure if the current `Signer` interface is compatible with this design though.

**Comment by Slamdunk** [comment]
I confirm that the current interfaces are enough to get it working without external deps.

The good news end here though: no PHP tool exist to help you with the low-level implementation, so the calculations have to be done hardcoded "by hand".

@Spomky has already did it nicely in his https://github.com/web-token/jwt-framework library:

1. Implementation: [src/SignatureAlgorithm/RSA/Util/RSA.php](https://github.com/web-token/jwt-framework/blob/v3.0.7/src/SignatureAlgorithm/RSA/Util/RSA.php)
2. Test with RFC link and dataset: [tests/SignatureAlgorithm/RSA/RSAPSSSignatureTest.php](https://github.com/web-token/jwt-framework/blob/v3.0.7/tests/SignatureAlgorithm/RSA/RSAPSSSignatureTest.php)

**Comment by Spomky** [comment]
Hi @Slamdunk,

Note that without GMP or BCMath, it will be very slow to compute or verify the signatures.

**Comment by Ocramius** [comment]
@Spomky make it work, make it fast in a second step :-)

**Comment by lcobucci** [comment]
@FractalizeR would you like to send a PR implementing it?

**Comment by SvenRtbg** [comment]
I was looking at this issue, investigating options, and I wouldn't mind some feedback.

So PS256 effectively is RS256 with a different padding scheme. However, PHP isn't currently supporting it, despite the openssl library possibly offering support (I suppose, because of https://bugs.php.net/bug.php?id=80495 which sounds like the PHP openssl extension would just be missing some parameters that can be passed).

So whatever implementation would make it happen has to do a userland implementation, and as has been stated, the algorithm for the padding (PS256 https://datatracker.ietf.org/doc/html/rfc7518#section-3.5 referencing https://datatracker.ietf.org/doc/html/rfc3447#section-8.1) requires treating a bytestring as a big integer and vice versa for the important algorithm steps.

There are implementations that do this work already. Simply doing BigInteger math is done by "brick/math", but if pulling in external packages, one might also opt for "phpseclib/phpseclib", which offers the entire hashing implementation ready-to-go. Both would cover utilizing extensions like GMP or BCMATH and fallback to a native implementation - no need to reinvent the wheel here.

However, even though PhpSecLib is actively maintained, version 2 allows PHP >=5.3.3, and version 3 (which has dependencies) asks for PHP >=5.6.1 - feels a bit awkward. At the same time, nobody questions the age of `openssl_sign()`. :)

So what would you be willing to do here? Add PS256 to this library at the cost of dealing with another dependency?

**Comment by lcobucci** [comment]
Hmmm, I see no reaction from PHP core developers on the bug (nor any mention of it on the indexed entries of https://externals.io/)... 

Although it's a niche need, having that ability at the language level might be interesting.
@Girgias would you have thoughts and/or tips on how to approach that?

Bringing that to userland is also okay (even with phpseclib)...
Sure, it comes at the expense of adding dependencies/extensions.
Having big integer math can also facilitate the processing of JWKs, so we might not have a way out of it in the end.

**Comment by Girgias** [comment]
For the OpenSSL thing, open an issue on php-src. I don't really know anything about OpenSSL nor do I maintain that extension, soooo there might be a good reason for no support, or it is just an oversight.

**Comment by SvenRtbg** [comment]
@FractalizeR If you want to give it a try, there's a (currently untagged) extension of lcobucci/jwt over at https://github.com/lcobucci/jwt-rsassa-pss - it's still missing all of the nice infrastructure stuff, so you'd need to add the repo explicitly and use "^0.1.0@dev" or similar.

**Comment by Spomky** [comment]
> ..., soooo there might be a good reason for no support, or it is just an oversight.

I am not a security expert, but I have read enough over the years to say that, yes, there are valid reasons to move away from RSA (non-exhaustive list):

* Shor’s algorithm on quantum computers can break RSA by factoring large integers (though [this is not yet feasible in practice](https://idtechwire.com/shanghai-university-researchers-claim-success-in-quantum-hack-of-rsa-encryption/)).
* RSA keys must keep growing to remain secure, which affects speed and resource usage ([the recommended key size is 2048 bits or more](https://csrc.nist.gov/pubs/sp/800/131/a/r2/final)).
* Elliptic curve and post-quantum algorithms provide stronger security with smaller keys.
* Implementation errors (in key generation, padding, etc.) and side-channel attacks can weaken RSA.
* Protocols (e.g., TLS 1.3, WireGuard) and agencies (NIST, ANSSI) increasingly favor elliptic curve and post-quantum solutions.

**Comment by SvenRtbg** [comment]
The 1.0.0 release was just created over there: https://github.com/lcobucci/jwt-rsassa-pss/releases/tag/1.0.0, effectively resolving the original issue.

**Comment by SvenRtbg** [comment]
@Spomky I agree RSA-related algorithms are quite dated, hence a matching comment in the readme:

> The RSASSA-PSS ([probabilistic signature scheme](https://en.wikipedia.org/wiki/Probabilistic_signature_scheme)) padding algorithm suggests it's security is mathematically proven to relate to the [RSA problem](https://en.wikipedia.org/wiki/RSA_problem).
> 
> However, this statement alone should not influence your judgement when asked to choose a signature algorithm for a JWT that you create. Consider using an [elliptic curve signature](https://www.scottbrady91.com/jose/jwts-which-signing-algorithm-should-i-use) instead. If however you are required to consume a token signed with a PS algorithm, you have no choice.
